docs/source/reference/envs.rst

```diff
@@ -169,7 +186,37 @@ one can simply call:
         >>> print(a)
         9.81
 
-It is also possible to reset some but not all of the environments:
+TorchRL uses a private ``"_reset"`` key to indicate to the environment which
+sub-environments should be reset.
+This allows to reset some but not all of the environments.
+Keep in mind that this is a private key, and it should only be used when coding
+specific environment features that are internal facing.
+The way an environment deals with the ``"_reset"`` keys is proper to it and
+designing an environment that behaves according to ``"_reset"`` inputs is the
+developer's responsibility. The following assumptions are made:
+
+- Each ``"_reset"`` is paired with a ``"done"`` (and/or ``"terminated"``,
+  ``"truncated"`` etc.). This means that the following structure is not
+  allowed: ``TensorDict({"done": done, "nested": {"_reset": reset}}, [])``.
+- A reset at one level precludes the presence of a ``"_reset"`` at lower
+  levels below it. For instance, this is not allowed:
+  ``TensorDict({"_reset": reset_main, "nested": {"_reset": reset_leaf}}, [])``
+- When calling :meth:`env.reset(tensordict)` with a partial ``"_reset"`` entry
+  that will reset some but not all the done sub-environments, the input data
+  should contain the data of the sub-environemtns that are __not__ being reset.
+  The reason for this constrain is that the output of the ``env._reset(data)``
+  can only be predicted for the entries that are reset. For the others, we
+  cannot know in advance if they will be meaningful or not. For instance,
+  one could perfectly just pad the values of the non-reset environments, in
+  which case the data will be meaningless and should be discarded.
+  :meth:`~torchrl.envs.EnvBase.reset` will merge the input tensordict with
+  the one resulting from reset and pad missing values with 0 (except for
+  ``done`` signals).
+  The other reason why this constrain is enforced is that usually,
+  setting ``"_reset"`` is achieved by internal methods that have access to
+  the data at the previous step, and use that data to call ``reset`` (after
+  passing it through :func:`~torchrl.envs.utils.step_mdp`). We can safely
+  assume that the data presented to ``reset`` is therefore complete.
```

Comment:
Why do we have to enforce this? In my view we can:

- if no td is passed as input to `reset()` just return what you get from `_reset()`
- if a td is passed as input with partial keys (e.g. maybe just the _reset keys), I would not want to have the output padded by default, but i would want to update the keys from the input using the ones from `_reset()` and, when the input keys are missing, directly copy the ones from `_reset`

Reply:
> if no td is passed as input to reset() just return what you get from _reset()

this is what happens

> if a td is passed as input with partial keys (e.g. maybe just the _reset keys), I would not want to have the output padded by default, but i would want to update the keys from the input using the ones from _reset() and, when the input keys are missing, directly copy the ones from _reset

that was the case so far but it's very tricky and brittle. Let me explain why:
- First I think we will all agree that we need a rule that applies to all envs, regardless of whether they are stateful, stateless, simple or nested / MARL. The rule must be transparent and always the same (in envs, collectors etc).
- Say we pass an input with just a `"_reset"` entry. We get an output from _reset. __Assuming__ the environment is stateful, it could make sense to say that all the data is valid. But that's only assuming the env is stateful __and__ that the users comply with the rule stated above (for which we have no guarantee). If it's stateless, we have no idea what the other data means. Also one could perfectly do
```python
def _reset(td):
    obs = torch.zeros(size)
    obs[td.get("_reset")] = some_obs_reset
    return TensorDict({"obs": obs}, [])
```
and IMO this should be valid (again, regardless of whether the env is stateful or stateless). However this will eventually break if we consider everything that comes out of _reset as valid.
- In batched envs, we have access to the buffers. We'll copy the entire buffer after a call to `_reset`, also for sub-envs that have not been reset. Then we make the wishful guess that the data of non-reset sub-envs is consistent with the data at the previous step, but we have no way to actually make sure that it's the case (eg, stateless envs and similar). These buffers should definitely __not__ be used to complete the data. 

In summary, choosing to "sometimes" use the data returned by the env has multiple ramifications and will cause a lot of headache long-term as it does not apply across all envs, and makes lots of unwated assumptions about internal machinery of the envs.  `EnvBase` responsibility is to provide a common working framework for environments given a set of rules. I feel that choosing to sometimes use or not the non-reset data coming from a reset is too opinionated for EnvBase to fullfill its purpose.

Reply:
I think you misunderstood my point.

> and IMO this should be valid (again, regardless of whether the env is stateful or stateless). However this will eventually break if we consider everything that comes out of _reset as valid.

I agree that this should be valid and I am _not_ proposing to consider everything that comes out of _reset as valid.

My point is about corner cases where reset is called with a partially complete input.
From the docstring it seems like this corner case is treated with padding of 0s.

My proposition is to treat it with data from _reset() instead of 0s, which are always guaranteed to be invalid.

Another option is to throw an error in the cases where we would pad

Reply:
in other words, the merge described in the docstring could happen without any padding, just creating a new tensordict from the 2 with the following rules

- when _reset is true use output of _reset()
- when _reset is false use data from  reset() input
- in case  _reset is false and there is no data from  reset() input, use data from _reset() (instead of 0s)

Reply:
I understood this point thanks.
That introduces inconsistencies between collectors (which call where and copy the previous "next" data) and calls to reset. We need a one fits all solution, and that doesn't make the cut.

Reply:
Let's imagine the following scenario with this: 
someone builds an app where they call `reset(data)` and use the non-reset data out of it.
Next they use the env in the collector and things break. Apparently there is not reason for it to fail since the collector should do the exact same thing.
We need to be super explicit regarding the non-safety of using data from reset that is not matched by a `"_reset"` entry. Allowing data to be passed is dangerous I think

Reply:
> That introduces inconsistencies between collectors (which call where and copy the previous "next" data) and calls to reset. We need a one fits all solution, and that doesn't make the cut.

So collectors pad with zeros?

I don't think this can be compared with collectors as this edge case I am talking about cannot happen within them as they always present complete data.

My only point here is that I don't see a case in which padding with zeros in `env.reset()` is beneficial to the user.
I was just trying to propose a solution to this but any solution works for me.

What I want to avoid is some 0s silently added to data in core environment calls

Reply:
> So collectors pad with zeros?

No they use where but they override the previous values. The point of padding is to "invalidate" data.

> My only point here is that I don't see a case in which padding with zeros in env.reset() is beneficial to the user.

It avoids letting people thing that data can be used if it's not reset, that's all. It's important because otherwise you can have silent bugs when using that data + inconsistencies if you're using it between collector and pure env execution. It's just a way of making the env robust and have a one-fits-all solution. 

The option of not padding is kind of allowing the data to be used, but if you swap your gym with an brax env in a parallel / serial execution, you may have polluted data when calling reset in the brax case and not in gym. To prevent that, we treat everything the same way. 

What's the case against padding precisely?

Reply:
I am not sure padding is a non-silent way to invalidate data.

In my opinion it will silently add arbitrary data and return it as if it was valid.

How can users know if padding happens?

> The option of not padding is kind of allowing the data to be used, but if you swap your gym with an brax env in a parallel / serial execution, you may have polluted data when calling reset in the brax case and not in gym. To prevent that, we treat everything the same way.

If treating everything the same way is padding, you are still polluting the data.

I would consider other alternatives, maybe error?

Reply:
> How can users know if padding happens?

They shouldn't care, the data is not valid. Why should they "know" if we pad, since sometimes the data will be rubbish anyway. The crucial point here is that we provide an API to code envs but we have no way of telling what happens within `_reset`, so we must be very transparent regarding what comes out of it for `"_reset"`s that are `False`. I think that "sometimes this work because my env is stateful and obbeys by gym standards" or anything like that is a no go: it makes the environments non-interchangeable.

This is what I'm trying to solve here: if we don't pad, we're treating different envs differently without consistency. Example:

```python
env = SerialEnv(2, lambda: GymEnv("Pendulum-v1"))
...
env.reset(TensorDict({"_reset": [True, False]}, [2])) # all data is valid
```

```python
env = BraxEnv(..., batch_size=[2])
...
env.reset(TensorDict({"_reset": [True, False]}, [2])) # we don't know anything about the second batch of data
```

Let me revert the question: how can the user know in which case the data can be used or not? From the outside, there is no way to tell, no attribute in env that can point you in that direction.

Add to that the inconsistency between this and collectors and we end up with a big stack of methods to call reset and return data that differ from each other and will be hard to track in the future. By having a clear pact (_reset = True means valid, _reset = False means 0) we are consistent.

I don't think we want an error either because one could have an env that takes care of returning the fully fleshed data after partial `_reset`.


Let's not forget a very important piece here: using `"_reset"` is reserved to TorchRL's components by definition. Anyone passing a tensordict with `"_reset"` in their code base (out of `pytorch/rl`) is doing it at their own risk. So the "user" shouldn't care about this. Developers should, but we document things in the doc so it's fine (they can know what is padded through the inverse of the `"_reset"` mask to answer the initial question).
This change is therefore not bc-breaking and safe. This is not public facing. And it will truly makes maintenance easier.

Reply:
> I don't think we want an error either because one could have an env that takes care of returning the fully fleshed data after partial _reset.

this is surprising since I thought this PR was here to forbid this. How would one be able to do this if we discard all data from `_reset()` that is not `_reset`?

```python
env = BraxEnv(..., batch_size=[2])
...
env.reset(TensorDict({"_reset": [True, False]}, [2])) # we don't know anything about the second batch of data
```
I still do not super understand why, if we don't know anything about the second batch of data, do we want to force 0s there. What is a user that is served with that output supposed to do? other data might be tangentially zero so even checking what is 0 would not tell what is valid. Plus `reset()` will remove the `_reset` flags, so there is really no way to understand whats valid.

The risk here is that all that data will be passed to another component and the 0s interpreted as real data.

Do you see what I am referring to? Something like
```python
policy(env.reset(TensorDict({"_reset": [True, False]}, [2])))
```
will silently use the zeros

Is there no solution where, whatever i pass to reset(), all its output will be all valid? 
maybe we can say that _reset() should return all valid output even with partial resets as a convention and then we are sure brax will do that

Reply:
> this is surprising since I thought this PR was here to forbid this. How would one be able to do this if we discard all data from _reset() that is not "_reset"?

fully fleshed doesn't mean valid, it just means contiguous (possibly masked or meaningless) data.
For internal usage it can be useful to call `env.reset` with a partial "_reset" without passing the full data as input, and since it's internal facing there's no need for an exception.

> What is a user that is served with that output supposed to do? 

the user never ever sees that, as stated earlier.

> other data might be tangentially zero so even checking what is 0 would not tell what is valid.

the developer has access to the `"_reset"` entry which masks the data, so one can very quickly know what is valid.

> Plus reset() will remove the _reset flags, so there is really no way to understand whats valid.

One still has the `"done"` from before and I think in all cases can backtrack what was being reset. 

```python
policy(env.reset(TensorDict({"_reset": [True, False]}, [2])))
```
that will never happen because `"_reset"` is just used within TorchRL. If  users are using it, they're not doing themselves a favour. `_` means private and should be considered as such. We don't have a single call to a policy like this in the code base I believe. And in the case of many envs it will still be very wrong to do it because the data of the second batch can or cannot be corrupted and there's not way to tell. 

 > Is there no solution where, whatever i pass to reset(), all its output will be all valid?

that can happen but since it cannot be always the case we cannot assume it and we must prevent that sort of usage at all cost!

Reply:
To rephrase: even without this PR, 
```python
policy(env.reset(TensorDict({"_reset": [True, False]}, [2])))
```
is wrong and should not be used (or at least not expecting that the policy is doing anything meaningful with index `1`)